File: 1/app/cogs/models/sequence.py
from typing import List
from pydantic import BaseModel, Field

class SequenceItemNote(BaseModel):
  pitch: str = Field(..., description="Musical pitch identifier")
  octave: int = Field(..., description="Octave")

class SequenceItem(BaseModel):
  br: bool = Field(False, description="Should line break after this note")
  notes: List[SequenceItemNote] = Field(..., description="List of notes in this sequence item")

class Sequence(BaseModel):
  items: List[SequenceItem] = Field(..., description="Note sequence as an ordered list of sequence items")

File: 2/src/types/sequence.ts
import type { NoteKey } from "./noteKey";

export interface Sequence {
  items: SequenceItem[];
}

export interface SequenceItem {
  br: boolean;
  notes: NoteKey[];
}

export interface SequenceCorrection {
  noteCount: number;
  sequenceLength: number;
}